File: src/1191-K-Concatenation-Maximum-Sum/1191.py
class Solution:
    def kConcatenationMaxSum(self, arr: List[int], k: int) -> int:
        mod = 10**9 + 7
        def Kadane(arr):
            res = cur = 0
            for num in arr:
                cur = max(cur, 0) + num
                res = max(res, cur)
            return res
        return ((k - 2) * max(sum(arr), 0) + Kadane(arr * 2)) % mod if k > 1 else Kadane(arr) % mod

File: src/1191-K-Concatenation-Maximum-Sum/1191.cpp
class Solution 
{
public:
    int kConcatenationMaxSum(vector<int>& arr, int k) 
    {
        int n = arr.size(), mod = 1e9 + 7, cur = 0, res = 0;
        long long total = accumulate(arr.begin(), arr.end(), 0)%mod;
        for (int i = 0; i < n*min(k, 2); ++i)
        {
            cur = max(cur, 0) + arr[i%n];
            res = max(res, cur);
        }
        return max({0, res, int(((k-2)*total + res)%mod)});
    }
};